npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,442 @@
+  able to send traffic to a logs-sink, hosted at `logs-storage.com`
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  any of the managed services provided by my Cloud Provider. Since the cloud
+  provider has a well known parent domain, I want to allow Pods to send traffic
+  to all sub-domains using a wild-card selector -- `*.my-cloud-provider.com`
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // This field acts as an exact match. It doesn't match any other
+    // domains of that level or even any subdomains.
+    // Examples:
+    //   - `kubernetes.io` matches only `kubernetes.io`.
+    //     It does not match "www.kubernetes.io", "blog.kubernetes.io",
+    //     "my-kubernetes.io", or "wikipedia.org".
+    //   - `blog.kubernetes.io` matches only "blog.kubernetes.io".
+    //     It does not match "www.kubernetes.io" or "kubernetes.io".
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])\.)*([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?$`
+    Name string `json:"names,omitempty"`
+
+    //  Pattern allows matching FQDNs using wildcard specifiers.
+    //
+    // "*" matches 0 or more DNS valid characters (except for "."), and may
+    // occur anywhere in the pattern.
+    //  Examples:
+    //    - `*.kubernetes.io` matches subdomains of kubernetes.io at that level.
```

Comment:
I asked my colleagues in our solutions team for examples of matching multiple URL segments, we do have customers that use * to match multiple parts in a prefix.  some examples they were able to share:

* `<team>.<platform>.example.com` matched with `*.example.com`
* `<sessionID>.<session>.<random>.<region>.cache.amazonaws.com` matched with `*.cache.amazonaws.com`

I think the AWS one would be `*.*.*.*.cache.amazonaws.com` in the proposed spec, which feels a bit messy.

Reply:
very nice examples, thank you for gathering!
I am getting more convinced every day that `*` should match any number of labels, unless someone can point out a security/performance/etc concerns about doing that

Reply:
I'm ok with that -- I can see CDNs getting pretty deeply nested so having `*` match multiple labels should be ok.

I don't imagine this should have any performance implications. Security-wise, you sacrifice some explicitness. But on the other hand, when you allow `*.domain.com`, you've already indicated you trust the owner `domain.com`

Reply:
In https://github.com/cilium/cilium/issues/22081, I've been discussing this topic with maintainers of Cilium (wich regards to `CiliumNetworkPolicy`) . Though still not decided or implemented, the dialog has been leaning towards:

  - `**` matches any character (including `.`). Allows wildcarding of multiple levels of subdomains.
  - `*` matches any character except `.`. Allows wildcarding of a single subdomain.

Perhaps this can be useful, if not just for some additional perspectives.